scalableminds/webknossos
# Zarr streaming

### URL of deployed dev instance (used for testing):
- https://zarrstreaming.webknossos.xyz/

### Steps to test:
- Use this script to test:
```
def _assert_check_equality(args) -> None:
    view_a, view_b, _ = args
    assert np.array_equal(view_a.read(absolute_offset=(3072, 3072, 512), size=(32, 32, 32)),
                          view_b.read(absolute_offset=(3072, 3072, 512), size=(32, 32, 32)))
    print("They are equal!!")


if __name__ == "__main__":
    p = UPath('http://localhost:9000/data/zarr/sample_organization/l4_sample/', **{"headers": {"X-Auth-Token": "secretSampleUserToken"}})
    assert p.exists()

    disk = wk.Dataset.open("/Users/leo/Documents/work/scm/webknossos/binaryData/sample_organization/l4_sample/")
    http = wk.Dataset.open(p)

    mag_disk = disk.get_layer("color").get_mag("1-1-1")
    mag_http = http.get_layer("color").get_mag("1-1-1")
    mag_disk.for_zipped_chunks(_assert_check_equality, mag_http)
```

There are currently 4 issues on the client side: 1) `is_dir` is wrongly implemented and mostly returns false in fsspec (should be fixed in [UPath PR](https://github.com/fsspec/universal_pathlib/pull/58/files#)), 2) `encoding=utf-8` caused a problem, 3) `iterdir` in UPath uses a false implementation of `_sub_path` (should be fixed in [UPath](https://github.com/fsspec/universal_pathlib/pull/58/files#)), and 4) `is_file` check, doesn't check first whether file exists which leads to an exception (change first line of  `wk.WKWArray.open` to `if (path / "header.wkw").exists() and (path / "header.wkw").is_file():`

### Issues:
- fixes #6106 

------
(Please delete unneeded items, merge only when none are left open)

------
**ToDo**
- [x] check with Zarr files
- [x] check with different mags
- [x] check with different layer names
